File: routes/product.js
// const express = require("express");
// const Product = require("../models/Product");
// const auth = require("../middleware/auth");

// const router = express.Router();

// // POST /api/products - Create a new product
// router.post("/", auth, async (req, res) => {
//   try {
//     const { name, marketplace, marketplaceProductId, imageurl } = req.body; // <--- Accept imageurl

//     // Basic validation
//     if (!name) {
//       return res.status(400).json({ message: "Product name is required" });
//     }

//     // Create product with current user as seller
//     const product = new Product({
//       name,
//       seller: req.user.id, // Get from JWT token
//       marketplace,
//       marketplaceProductId,
//       campaigns: [],
//       imageurl, // <--- Save imageurl if present
//     });

//     await product.save();

//     // Populate seller info before sending response
//     await product.populate("seller", "name email");

//     res.status(201).json({
//       message: "Product created successfully",
//       product,
//     });
//   } catch (error) {
//     console.error("Create product error:", error);
//     res.status(500).json({ message: "Server error" });
//   }
// });

// // GET /api/products - Get all products for the current user
// router.get("/", auth, async (req, res) => {
//   try {
//     const products = await Product.find({ seller: req.user.id })
//       .populate("seller", "name email")
//       .populate("campaigns", "name status")
//       .sort({ createdAt: -1 });

//     res.json({
//       count: products.length,
//       products,
//     });
//   } catch (error) {
//     console.error("Get products error:", error);
//     res.status(500).json({ message: "Server error" });
//   }
// });

// // GET /api/products/:id - Get a specific product by ID
// router.get("/:id", auth, async (req, res) => {
//   try {
//     const product = await Product.findOne({
//       _id: req.params.id,
//       seller: req.user.id, // Ensure user can only access their own products
//     })
//       .populate("seller", "name email")
//       .populate("campaigns", "name status qrCodeUrl");

//     if (!product) {
//       return res.status(404).json({ message: "Product not found" });
//     }

//     res.json(product);
//   } catch (error) {
//     console.error("Get product error:", error);
//     if (error.name === "CastError") {
//       return res.status(400).json({ message: "Invalid product ID" });
//     }
//     res.status(500).json({ message: "Server error" });
//   }
// });

// // PUT /api/products/:id - Update a product
// router.put("/:id", auth, async (req, res) => {
//   try {
//     const { name, marketplace, marketplaceProductId, imageurl } = req.body; // <--- Accept imageurl

//     // Find product and ensure it belongs to the current user
//     const product = await Product.findOne({
//       _id: req.params.id,
//       seller: req.user.id,
//     });

//     if (!product) {
//       return res.status(404).json({ message: "Product not found" });
//     }

//     // Update fields
//     if (name !== undefined) product.name = name;
//     if (marketplace !== undefined) product.marketplace = marketplace;
//     if (marketplaceProductId !== undefined)
//       product.marketplaceProductId = marketplaceProductId;
//     if (imageurl !== undefined) product.imageurl = imageurl; // <--- Update imageurl

//     await product.save();

//     // Populate before sending response
//     await product.populate("seller", "name email");
//     await product.populate("campaigns", "name status");

//     res.json({
//       message: "Product updated successfully",
//       product,
//     });
//   } catch (error) {
//     console.error("Update product error:", error);
//     if (error.name === "CastError") {
//       return res.status(400).json({ message: "Invalid product ID" });
//     }
//     res.status(500).json({ message: "Server error" });
//   }
// });

// // DELETE /api/products/:id - Delete a product
// router.delete("/:id", auth, async (req, res) => {
//   try {
//     // Find product and ensure it belongs to the current user
//     const product = await Product.findOne({
//       _id: req.params.id,
//       seller: req.user.id,
//     });

//     if (!product) {
//       return res.status(404).json({ message: "Product not found" });
//     }

//     // Check if product has active campaigns
//     if (product.campaigns && product.campaigns.length > 0) {
//       return res.status(400).json({
//         message:
//           "Cannot delete product with active campaigns. Please delete campaigns first.",
//       });
//     }

//     await Product.findByIdAndDelete(req.params.id);

//     res.json({ message: "Product deleted successfully" });
//   } catch (error) {
//     console.error("Delete product error:", error);
//     if (error.name === "CastError") {
//       return res.status(400).json({ message: "Invalid product ID" });
//     }
//     res.status(500).json({ message: "Server error" });
//   }
// });

// module.exports = router;
const express = require("express");
const Product = require("../models/Product");
const auth = require("../middleware/auth");

const router = express.Router();

// POST /api/products - Create a new product
router.post("/", auth, async (req, res) => {
  try {
    const { name, amazonAsin, flipkartFsn, imageurl } = req.body;

    // Basic validation
    if (!name) {
      return res.status(400).json({ message: "Product name is required" });
    }

    // Validate at least one marketplace is provided
    if (!amazonAsin && !flipkartFsn) {
      return res.status(400).json({ 
        message: "At least one marketplace is required (Amazon ASIN or Flipkart FSN)" 
      });
    }

    // Validate ASIN format (Amazon ASINs are typically 10 characters)
    if (amazonAsin && (typeof amazonAsin !== 'string' || amazonAsin.length !== 10)) {
      return res.status(400).json({ 
        message: "Amazon ASIN must be a 10-character string" 
      });
    }

    // Validate FSN format (Flipkart FSNs are typically alphanumeric)
    if (flipkartFsn && (typeof flipkartFsn !== 'string' || flipkartFsn.length < 5)) {
      return res.status(400).json({ 
        message: "Flipkart FSN must be a valid string (minimum 5 characters)" 
      });
    }

    // Create product with current user as seller
    const product = new Product({
      name,
      seller: req.user.id,
      amazonAsin,
      flipkartFsn,
      campaigns: [],
      imageurl,
    });

    await product.save();

    // Populate seller info before sending response
    await product.populate("seller", "name email");

    res.status(201).json({
      message: "Product created successfully",
      product,
    });
  } catch (error) {
    console.error("Create product error:", error);
    
    // Handle validation errors
    if (error.name === 'ValidationError') {
      return res.status(400).json({ message: error.message });
    }
    
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/products - Get all products for the current user
router.get("/", auth, async (req, res) => {
  try {
    const products = await Product.find({ seller: req.user.id })
      .populate("seller", "name email")
      .populate("campaigns", "name status")
      .sort({ createdAt: -1 });

    res.json({
      count: products.length,
      products,
    });
  } catch (error) {
    console.error("Get products error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/products/:id - Get a specific product by ID
router.get("/:id", auth, async (req, res) => {
  try {
    const product = await Product.findOne({
      _id: req.params.id,
      seller: req.user.id,
    })
      .populate("seller", "name email")
      .populate("campaigns", "name status qrCodeUrl");

    if (!product) {
      return res.status(404).json({ message: "Product not found" });
    }

    res.json(product);
  } catch (error) {
    console.error("Get product error:", error);
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid product ID" });
    }
    res.status(500).json({ message: "Server error" });
  }
});

// PUT /api/products/:id - Update a product
router.put("/:id", auth, async (req, res) => {
  try {
    const { name, amazonAsin, flipkartFsn, imageurl } = req.body;

    // Find product and ensure it belongs to the current user
    const product = await Product.findOne({
      _id: req.params.id,
      seller: req.user.id,
    });

    if (!product) {
      return res.status(404).json({ message: "Product not found" });
    }

    // Update fields
    if (name !== undefined) product.name = name;
    if (amazonAsin !== undefined) product.amazonAsin = amazonAsin;
    if (flipkartFsn !== undefined) product.flipkartFsn = flipkartFsn;
    if (imageurl !== undefined) product.imageurl = imageurl;

    // Validate at least one marketplace after update
    if (!product.amazonAsin && !product.flipkartFsn) {
      return res.status(400).json({ 
        message: "At least one marketplace is required (Amazon ASIN or Flipkart FSN)" 
      });
    }

    // Validate ASIN format if provided
    if (product.amazonAsin && (typeof product.amazonAsin !== 'string' || product.amazonAsin.length !== 10)) {
      return res.status(400).json({ 
        message: "Amazon ASIN must be a 10-character string" 
      });
    }

    // Validate FSN format if provided
    if (product.flipkartFsn && (typeof product.flipkartFsn !== 'string' || product.flipkartFsn.length < 5)) {
      return res.status(400).json({ 
        message: "Flipkart FSN must be a valid string (minimum 5 characters)" 
      });
    }

    await product.save();

    // Populate before sending response
    await product.populate("seller", "name email");
    await product.populate("campaigns", "name status");

    res.json({
      message: "Product updated successfully",
      product,
    });
  } catch (error) {
    console.error("Update product error:", error);
    
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid product ID" });
    }
    if (error.name === 'ValidationError') {
      return res.status(400).json({ message: error.message });
    }
    
    res.status(500).json({ message: "Server error" });
  }
});

// DELETE /api/products/:id - Delete a product
router.delete("/:id", auth, async (req, res) => {
  try {
    // Find product and ensure it belongs to the current user
    const product = await Product.findOne({
      _id: req.params.id,
      seller: req.user.id,
    });

    if (!product) {
      return res.status(404).json({ message: "Product not found" });
    }

    // Check if product has active campaigns
    if (product.campaigns && product.campaigns.length > 0) {
      return res.status(400).json({
        message:
          "Cannot delete product with active campaigns. Please delete campaigns first.",
      });
    }

    await Product.findByIdAndDelete(req.params.id);

    res.json({ message: "Product deleted successfully" });
  } catch (error) {
    console.error("Delete product error:", error);
    if (error.name === "CastError") {
      return res.status(400).json({ message: "Invalid product ID" });
    }
    res.status(500).json({ message: "Server error" });
  }
});

// GET /api/products/marketplace/:marketplace - Get products by marketplace
router.get("/marketplace/:marketplace", auth, async (req, res) => {
  try {
    const { marketplace } = req.params;
    let query = { seller: req.user.id };

    if (marketplace.toLowerCase() === 'amazon') {
      query.amazonAsin = { $exists: true, $ne: null };
    } else if (marketplace.toLowerCase() === 'flipkart') {
      query.flipkartFsn = { $exists: true, $ne: null };
    } else {
      return res.status(400).json({ message: "Invalid marketplace. Use 'amazon' or 'flipkart'" });
    }

    const products = await Product.find(query)
      .populate("seller", "name email")
      .populate("campaigns", "name status")
      .sort({ createdAt: -1 });

    res.json({
      marketplace,
      count: products.length,
      products,
    });
  } catch (error) {
    console.error("Get products by marketplace error:", error);
    res.status(500).json({ message: "Server error" });
  }
});

// In your products routes file
router.post("/bulk", auth, async (req, res) => {
  try {
    const products = req.body.products;
    if (!Array.isArray(products) || products.length === 0)
      return res.status(400).json({ message: "No products found in upload." });

    // Further validation can be done here
    const docs = products.map(({ name, flipkartFsn, amazonAsin, imageurl }) => ({
      name,
      flipkartFsn,
      amazonAsin,
      imageurl,
      seller: req.user.id,
    }));
    await Product.insertMany(docs);
    res.json({ message: "Bulk products uploaded successfully." });
  } catch (err) {
    res.status(500).json({ message: err.message || "Bulk upload failed." });
  }
});

module.exports = router;

File: yourapp-go/controllers/product.go
package controllers

import (
	"context"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"yourapp/config"
	"yourapp/middleware"
	"yourapp/models"
)

// CreateProductRequest represents the request body for creating a product
type CreateProductRequest struct {
	Name        string `json:"name" binding:"required"`
	AmazonASIN  string `json:"amazonAsin,omitempty"`
	FlipkartFSN string `json:"flipkartFsn,omitempty"`
	ImageURL    string `json:"imageurl,omitempty"`
}

// UpdateProductRequest represents the request body for updating a product
type UpdateProductRequest struct {
	Name        *string `json:"name,omitempty"`
	AmazonASIN  *string `json:"amazonAsin,omitempty"`
	FlipkartFSN *string `json:"flipkartFsn,omitempty"`
	ImageURL    *string `json:"imageurl,omitempty"` // Pointer to distinguish between empty and not provided
	

}

// BulkProductRequest represents the request for bulk product upload
type BulkProductRequest struct {
	Products []CreateProductRequest `json:"products" binding:"required,min=1"`
}

// CreateProduct creates a new product
func CreateProduct(c *gin.Context) {
	var req CreateProductRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		log.Printf("Validation error: %v", err)
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Invalid request data",
		})
		return
	}

	// Validate at least one marketplace is provided
	if req.AmazonASIN == "" && req.FlipkartFSN == "" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "At least one marketplace is required (Amazon ASIN or Flipkart FSN)",
		})
		return
	}

	// Validate ASIN format (Amazon ASINs are typically 10 characters)
	if req.AmazonASIN != "" && len(req.AmazonASIN) != 10 {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Amazon ASIN must be a 10-character string",
		})
		return
	}

	// Validate FSN format (Flipkart FSNs are typically alphanumeric, minimum 5 characters)
	if req.FlipkartFSN != "" && len(req.FlipkartFSN) < 5 {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Flipkart FSN must be a valid string (minimum 5 characters)",
		})
		return
	}

	// Get user ID from context
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Create new product
	product := models.NewProduct()
	product.Name = req.Name
	product.Seller = userID
	product.AmazonASIN = req.AmazonASIN
	product.FlipkartFSN = req.FlipkartFSN
	product.ImageURL = req.ImageURL

	// Insert product
	result, err := db.Collection(product.CollectionName()).InsertOne(ctx, product)
	if err != nil {
		log.Printf("Database insert error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to create product"})
		return
	}

	product.ID = result.InsertedID.(primitive.ObjectID)

	c.JSON(http.StatusCreated, gin.H{
		"message": "Product created successfully",
		"product": product.ToResponse(),
	})
}

// GetProducts retrieves all products for the current user
func GetProducts(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Find products with sorting
	findOptions := options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}})

	cursor, err := db.Collection("products").Find(ctx, bson.M{"seller": userID}, findOptions)
	if err != nil {
		log.Printf("Database error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		return
	}
	defer cursor.Close(ctx)

	var products []models.Product
	if err := cursor.All(ctx, &products); err != nil {
		log.Printf("Error decoding products: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error decoding products"})
		return
	}

	// Initialize empty array if nil
	if products == nil {
		products = []models.Product{}
	}

	// Convert to response format
	productResponses := make([]*models.ProductResponse, len(products))
	for i, p := range products {
		productResponses[i] = p.ToResponse()
	}

	c.JSON(http.StatusOK, gin.H{
		"count":    len(products),
		"products": productResponses,
	})
}

// GetProductByID retrieves a specific product by ID
func GetProductByID(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	productID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	var product models.Product
	err = db.Collection("products").FindOne(ctx, bson.M{
		"_id":    productID,
		"seller": userID,
	}).Decode(&product)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Product not found"})
		} else {
			log.Printf("Database error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	c.JSON(http.StatusOK, product.ToResponse())
}

// UpdateProduct updates a product
func UpdateProduct(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	productID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID"})
		return
	}

	var req UpdateProductRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid request data"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Get current product
	var product models.Product
	err = db.Collection("products").FindOne(ctx, bson.M{
		"_id":    productID,
		"seller": userID,
	}).Decode(&product)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Product not found"})
		} else {
			log.Printf("Database error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	// Build update document
	update := bson.M{"$set": bson.M{}}

	if req.Name != nil {
		product.Name = *req.Name
		update["$set"].(bson.M)["name"] = *req.Name
	}
	if req.AmazonASIN != nil {
		product.AmazonASIN = *req.AmazonASIN
		update["$set"].(bson.M)["amazonAsin"] = *req.AmazonASIN
	}
	if req.FlipkartFSN != nil {
		product.FlipkartFSN = *req.FlipkartFSN
		update["$set"].(bson.M)["flipkartFsn"] = *req.FlipkartFSN
	}
	if req.ImageURL != nil {
		product.ImageURL = *req.ImageURL
		update["$set"].(bson.M)["imageurl"] = *req.ImageURL
	}

	// Validate at least one marketplace after update
	if err := product.Validate(); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": err.Error()})
		return
	}

	// Validate ASIN format if provided
	if product.AmazonASIN != "" && len(product.AmazonASIN) != 10 {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Amazon ASIN must be a 10-character string",
		})
		return
	}

	// Validate FSN format if provided
	if product.FlipkartFSN != "" && len(product.FlipkartFSN) < 5 {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Flipkart FSN must be a valid string (minimum 5 characters)",
		})
		return
	}

	// Update product
	if len(update["$set"].(bson.M)) > 0 {
		_, err = db.Collection("products").UpdateOne(
			ctx,
			bson.M{"_id": productID, "seller": userID},
			update,
		)

		if err != nil {
			log.Printf("Database update error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to update product"})
			return
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Product updated successfully",
		"product": product.ToResponse(),
	})
}

// DeleteProduct deletes a product
func DeleteProduct(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	productID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Get current product
	var product models.Product
	err = db.Collection("products").FindOne(ctx, bson.M{
		"_id":    productID,
		"seller": userID,
	}).Decode(&product)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Product not found"})
		} else {
			log.Printf("Database error: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		}
		return
	}

	// Check if product has active campaigns
	if product.Campaigns != nil && len(product.Campaigns) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Cannot delete product with active campaigns. Please delete campaigns first.",
		})
		return
	}

	// Delete product
	_, err = db.Collection("products").DeleteOne(ctx, bson.M{"_id": productID})
	if err != nil {
		log.Printf("Database delete error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to delete product"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Product deleted successfully"})
}

// GetProductsByMarketplace retrieves products filtered by marketplace
func GetProductsByMarketplace(c *gin.Context) {
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	marketplace := strings.ToLower(c.Param("marketplace"))

	// Validate marketplace
	if marketplace != "amazon" && marketplace != "flipkart" {
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "Invalid marketplace. Use 'amazon' or 'flipkart'",
		})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Build query based on marketplace
	query := bson.M{"seller": userID}
	if marketplace == "amazon" {
		query["amazonAsin"] = bson.M{"$exists": true, "$ne": ""}
	} else if marketplace == "flipkart" {
		query["flipkartFsn"] = bson.M{"$exists": true, "$ne": ""}
	}

	// Find products with sorting
	findOptions := options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}})

	cursor, err := db.Collection("products").Find(ctx, query, findOptions)
	if err != nil {
		log.Printf("Database error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error"})
		return
	}
	defer cursor.Close(ctx)

	var products []models.Product
	if err := cursor.All(ctx, &products); err != nil {
		log.Printf("Error decoding products: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error decoding products"})
		return
	}

	// Initialize empty array if nil
	if products == nil {
		products = []models.Product{}
	}

	// Convert to response format
	productResponses := make([]*models.ProductResponse, len(products))
	for i, p := range products {
		productResponses[i] = p.ToResponse()
	}

	c.JSON(http.StatusOK, gin.H{
		"marketplace": marketplace,
		"count":       len(products),
		"products":    productResponses,
	})
}

// BulkCreateProducts creates multiple products at once
func BulkCreateProducts(c *gin.Context) {
	var req BulkProductRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		log.Printf("Validation error: %v", err)
		c.JSON(http.StatusBadRequest, gin.H{
			"message": "No products found in upload.",
		})
		return
	}

	// Get user ID from context
	userID, exists := middleware.GetUserID(c)
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	db := config.GetMongoDB()
	if db == nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Database connection error"})
		return
	}

	// Prepare products for bulk insert
	docs := make([]interface{}, len(req.Products))
	for i, p := range req.Products {
		product := models.NewProduct()
		product.Name = p.Name
		product.Seller = userID
		product.AmazonASIN = p.AmazonASIN
		product.FlipkartFSN = p.FlipkartFSN
		product.ImageURL = p.ImageURL

		// Validate each product
		if err := product.Validate(); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"message": "Product at index " + string(rune(i)) + ": " + err.Error(),
			})
			return
		}

		docs[i] = product
	}

	// Bulk insert
	result, err := db.Collection("products").InsertMany(ctx, docs)
	if err != nil {
		log.Printf("Bulk insert error: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"message": "Bulk upload failed.",
		})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"message":        "Bulk products uploaded successfully.",
		"insertedCount":  len(result.InsertedIDs),
	})
}
